### PR TITLE
Added abstract JdbcDoFn to execute lookup

### PR DESCRIPTION
**Pull request by shnapz**
More convenient way to implement lookup to Jdbc source. What can be added more:
syntax sugar for SCollection to integrate lookup lambda?

### CODE REVIEW

File: scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * A {@link DoFn} that performs synchronous lookup using JDBC connection.
+ *
+ * @param <A> input element type.
+ * @param <B> JDBC lookup value type.
+ */
+public abstract class JdbcDoFn<A, B> extends DoFn<A, B> {
+
+  private final SerializableFunction<Void, DataSource> dataSourceProviderFn;
```
Comment (RustedBones) [inline scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java:34]: I think `Supplier` is more adapted here

File: scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  private DataSource dataSource;
+  private Connection connection;
```
Comment (RustedBones) [inline scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java:36]: Looks those are interfaces without serialization guarantees. Probably worth to mark them as transient.

File: scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java (hunk 1 above)
Comment (shnapz) [reply scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java:34]: Neither of available `Supplier` interfaces are serializable

File: scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java (hunk 2 above)
Comment (shnapz) [reply scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java]: Done, also added unit test to verify serializable

File: scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java (hunk 1 above)
Comment (RustedBones) [reply scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java:34]: ha ok. Maybe we can simply add a 'nicer' API with the Supplier as constructor parameter, and convert it to the serializable function.

**Review by RustedBones — APPROVED** [review]

**Pull request merged**